lib/Sema/CSSimplify.cpp

```diff
@@ -5320,7 +5390,8 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
       if (kind >= ConstraintKind::Subtype &&
           nominal1->getDecl() != nominal2->getDecl() &&
           ((nominal1->isCGFloatType() || nominal2->isCGFloatType()) &&
-           (nominal1->isDouble() || nominal2->isDouble()))) {
+           (nominal1->isDouble() || nominal2->isDouble())) ||
+          useRestrictions && implicitConversionAvailable(type1, type2)) {
```

Comment:
Please extract this into a separate check, no need to try to plug it into the Double/CGFloat. As a side note - if you are working on a more general mechanism then I think we need to remove Double/CGFloat specific logic all together from here because it would be possible to declare a `init(implicit: CGFloat)` on `Double` and reverse one of `CGFloat` and use new infrastructure.